### DISCUSSION TITLE
Recommended Way to Handle Custom LSP Messages Sent from Client

### DISCUSSION THREAD

**Discussion by ballcoach12** [eclipse-langium/langium#504] [Q&A] [answered]
I am working on extending LSP to add my own custom messages for a particular domain.  One scenario is that, on the client, a user-facing command is invoked; when it is handled on the client, an LSP request is sent to the server, which will require the server to take some action.  This could also be done using the LSP's Command feature.  The server then sends either a progress notification or another type of notification back to the client in order to satisfy the requirements of LSP.  

Currently, I have configured my language project to have its own `language-server.ts` file that is similar to the one that ships with the Langium library, except that is has custom handlers in it for commands that it expects to receive from a client.  However, it seems like the scenario could also be satisfied through dependency injection or some sort of middleware.

Is there a recommended way to achieve this scenario besides the way that I am doing it?

**Comment by spoenemann** [comment, marked as answer]
You can add handlers to the `Connection` even after calling `startLanguageServer`. The only concern with that might be that you could miss custom messages that are received right after starting to listen on the connection. I already thought about adding an optional parameter for custom message handlers to that function.

**Reply by spoenemann** [reply]
See this example of a handler added after the server has started listening:

https://github.com/eclipse/sprotty-vscode/blob/c11b093fa90663c351ba2b66786172830fa6b4b0/examples/states-langium/language-server/src/main.ts#L29-L30
